x-pack/plugins/enterprise_search/public/applications/shared/unsaved_changes_prompt/unsaved_changes_prompt.tsx

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useEffect } from 'react';
+import { Prompt } from 'react-router-dom';
+
+import { i18n } from '@kbn/i18n';
+
+const DEFAULT_MESSAGE_TEXT = i18n.translate('xpack.enterpriseSearch.shared.unsavedChangesMessage', {
+  defaultMessage: 'Your changes have not been saved. Are you sure you want to leave?',
+});
+interface Props {
+  hasUnsavedChanges: boolean;
+  messageText?: string;
+}
+
+export const UnsavedChangesPrompt: React.FC<Props> = ({
+  hasUnsavedChanges,
+  messageText = DEFAULT_MESSAGE_TEXT,
+}) => {
+  useEffect(() => {
+    const handler = (event: BeforeUnloadEvent) => {
+      if (hasUnsavedChanges) {
+        // These 2 lines of code are the recommendation from MDN for triggering a browser prompt for confirming
+        // whether or not a user wants to leave the current site.
+        event.preventDefault();
+        event.returnValue = '';
+      }
+    };
+    // Adding this handler will prompt users if they are navigating to a new page, outside of the Kibana SPA
+    window.addEventListener('beforeunload', handler);
+    return () => window.removeEventListener('beforeunload', handler);
```

Comment:
Just checking my understanding, since your passing `[hasUnsavedChanges]` as the second argument to `useEffect`, this line will be run when `hasUnsavedChanges` changes, but before function to add the new listener will be called with the new value.  I'm always a little shaky on useEffect lifecycle, but I want to make sure here because calling `removeEventListener` is very very important so we never have multiple of these on the page.

Reply:
Super good question, I was curious as well so I looked it up:

> When exactly does React clean up an effect? React performs the cleanup when the component unmounts. However, as we learned earlier, effects run for every render and not just once. **This is why React also cleans up effects from the previous render before running the effects next time.** We’ll discuss why this helps avoid bugs and how to opt out of this behavior in case it creates performance issues later below.

Source: https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1 (bolding/emphasis mine)

I'm pretty sure this means that the return cleanup/removeEventListener handler is correctly running every render. This appears to be borne out by Jason's mounted unit tests as well.

Reply:
Yep, the tests cover this scenario pretty thoroughly.